assets/js/googlesitekit/datastore/user/key-metrics.js

```diff
@@ -257,6 +257,7 @@ const baseSelectors = {
 	 *
 	 * @return {Array<string>|undefined} An array of Key Metric widget slugs, or undefined if the user input settings are not loaded.
 	 */
+	/* eslint-disable complexity */
```

Comment:
Let's try to refactor this selector to reduce its complexity instead of disabling the eslint rule.

Reply:
@eugene-manuilov I am going to challenge this one, as I feel we'd be satisfying an eslint rule but worsening the developer experience here, having to constantly check a function located elsewhere just to read (or add to) the list of metrics mapped to answers. This is a perfectly readable array, and I think we can keep the ignore rule. WDYT?

Reply:
@10upsimon, here are perfectly readable arrays in two new selectors:

```js

	getRegularKeyMetricsWidgetIDs: createRegistrySelector( ( select ) => () => {
		const postTypes = select( CORE_SITE ).getPostTypes();
		const hasProductPostType = postTypes.some(
			( { slug } ) => slug === 'product'
		);

		return {
			publish_blog: [
				KM_ANALYTICS_RETURNING_VISITORS,
				KM_ANALYTICS_NEW_VISITORS,
				KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
				KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
			],
			publish_news: [
				KM_ANALYTICS_PAGES_PER_VISIT,
				KM_ANALYTICS_VISIT_LENGTH,
				KM_ANALYTICS_VISITS_PER_VISITOR,
				KM_ANALYTICS_MOST_ENGAGING_PAGES,
			],
			monetize_content: [
				KM_ANALYTICS_POPULAR_CONTENT,
				KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				KM_ANALYTICS_NEW_VISITORS,
				KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
			],
			sell_products_or_service: [
				hasProductPostType
					? KM_ANALYTICS_POPULAR_PRODUCTS
					: KM_ANALYTICS_POPULAR_CONTENT,
				KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
			],
			sell_products: [
				hasProductPostType
					? KM_ANALYTICS_POPULAR_PRODUCTS
					: KM_ANALYTICS_POPULAR_CONTENT,
				KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
			],
			provide_services: [
				KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
				KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				KM_ANALYTICS_POPULAR_CONTENT,
				KM_ANALYTICS_TOP_RETURNING_VISITOR_PAGES,
			],
			share_portfolio: [
				KM_ANALYTICS_NEW_VISITORS,
				KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
				KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
				KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
			],
		};
	} ),

	getConversionTailoredKeyMetricsWidgetIDs: createRegistrySelector(
		( select ) => () => {
			const postTypes = select( CORE_SITE ).getPostTypes();
			const hasProductPostType = postTypes.some(
				( { slug } ) => slug === 'product'
			);

			return {
				publish_blog: [
					KM_ANALYTICS_TOP_CATEGORIES,
					KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE,
					KM_ANALYTICS_TOP_RETURNING_VISITOR_PAGES,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
					KM_ANALYTICS_TOP_RECENT_TRENDING_PAGES,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
					KM_ANALYTICS_TOP_PAGES_DRIVING_LEADS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_LEADS,
				],
				publish_news: [
					KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
					KM_ANALYTICS_POPULAR_AUTHORS,
					KM_ANALYTICS_TOP_CITIES,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
					KM_ANALYTICS_TOP_RECENT_TRENDING_PAGES,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
					KM_ANALYTICS_TOP_PAGES_DRIVING_LEADS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_LEADS,
				],
				monetize_content: [
					KM_ANALYTICS_MOST_ENGAGING_PAGES,
					KM_ANALYTICS_POPULAR_CONTENT,
					KM_ANALYTICS_NEW_VISITORS,
					KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT,
					KM_ANALYTICS_VISIT_LENGTH,
					KM_ANALYTICS_VISITS_PER_VISITOR,
					KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				],
				sell_products_or_service: [
					hasProductPostType
						? KM_ANALYTICS_POPULAR_PRODUCTS
						: KM_ANALYTICS_POPULAR_CONTENT,
					KM_ANALYTICS_TOP_CITIES_DRIVING_PURCHASES,
					KM_ANALYTICS_TOP_DEVICE_DRIVING_PURCHASES,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_ADD_TO_CART,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_PURCHASES,
					KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT,
					KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				],
				sell_products: [
					hasProductPostType
						? KM_ANALYTICS_POPULAR_PRODUCTS
						: KM_ANALYTICS_POPULAR_CONTENT,
					KM_ANALYTICS_TOP_CITIES_DRIVING_PURCHASES,
					KM_ANALYTICS_TOP_DEVICE_DRIVING_PURCHASES,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_ADD_TO_CART,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_PURCHASES,
					KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT,
					KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				],
				provide_services: [
					KM_ANALYTICS_TOP_CITIES_DRIVING_LEADS,
					KM_ANALYTICS_TOP_PAGES_DRIVING_LEADS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_LEADS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE,
					KM_ANALYTICS_ENGAGED_TRAFFIC_SOURCE,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
					KM_ANALYTICS_POPULAR_CONTENT,
					KM_ANALYTICS_TOP_RETURNING_VISITOR_PAGES,
				],
				share_portfolio: [
					KM_ANALYTICS_TOP_CITIES_DRIVING_LEADS,
					KM_ANALYTICS_TOP_CONVERTING_TRAFFIC_SOURCE,
					KM_ANALYTICS_TOP_RETURNING_VISITOR_PAGES,
					KM_ANALYTICS_POPULAR_AUTHORS,
					KM_ANALYTICS_TOP_PAGES_DRIVING_LEADS,
					KM_ANALYTICS_TOP_TRAFFIC_SOURCE_DRIVING_LEADS,
					KM_ANALYTICS_POPULAR_CONTENT,
					KM_SEARCH_CONSOLE_POPULAR_KEYWORDS,
				],
			};
		}
	),
```

Then the `getAnswerBasedMetrics` selector becomes pretty simple:

```js
	getAnswerBasedMetrics: createRegistrySelector(
		( select ) => ( state, purposeOverride ) => {
			const userInputSettings =
				select( CORE_USER ).getUserInputSettings();

			if ( userInputSettings === undefined ) {
				return undefined;
			}

			const purpose =
				purposeOverride ?? userInputSettings?.purpose?.values?.[ 0 ];

			const showConversionTailoredMetrics =
				select( CORE_USER ).showConversionTailoredMetrics();

			const widgetIDs = showConversionTailoredMetrics
				? select( CORE_USER ).getRegularKeyMetricsWidgetIDs()
				: select(
						CORE_USER
				  ).getConversionTailoredKeyMetricsWidgetIDs();

			return widgetIDs[ purpose ] || [];
		}
	),
```

This is much easier to read and maintain without a need to disable the eslint rule.